### ISSUE TITLE
Bogus warning with `ItemsView`

### ISSUE THREAD

**Issue by OfekShilon** [python/mypy#19495]
**Bug Report**

Changing the type hint of an argument from `Iterable` to `ItemsView` causes a typing error -

> Argument 1 to "map" has incompatible type "Callable[[tuple[Callable[..., T], *tuple[Any, ...]]], T]"; expected "Callable[[tuple[object, ...]], T]"  [arg-type]

(A clear and concise description of what the bug is.)

**To Reproduce**

```python
from typing import Iterable, ItemsView, TypeVar, Any
from collections.abc import Callable

T = TypeVar('T')

def my_map(func: Callable[..., T], iterable: ItemsView[Any, Any], *args: object) -> list[T]:
    def _wrapped_func(func_and_args: tuple[Callable[..., T], *tuple[Any, ...]]) -> T:
        func = func_and_args[0]
        args = func_and_args[1:]
        return func(*args)

    func_and_args = [(func, it_arg, *args) for it_arg in iterable]      # <=======
    # error: Argument 1 to "map" has incompatible type "Callable[[tuple[Callable[..., T], *tuple[Any, ...]]], T]"; 
    #                                         expected "Callable[[tuple[object, ...]], T]"  [arg-type]

    result: list[T] = list(map(_wrapped_func, func_and_args))
    return result
```
[mypy playground link](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCSMApiAIYBGANsQDSEkQDOAaksQO70AqixLZEPQCCKOAFgAUKEhQAxmCo05MJGBRMAdJTmZsufAGEySyjSlTuUALxReCfoIAUAcm4uAlBckATYsCgIOAB9CDIEYLQwMB8nYABXFDkALihjU2piAG1NXJ4AXXokEnJM1KJSM2zROEKoACpBNCZUsAoAK2IVDygAWgA%2BKCokJhgs7nzkqSgZqD8A4I5yBAcfYISkuMS5YLIUNaaW2HiEGiz0qiqcvLs6%2BpgTs5r6XM18-J6BuynJWb%2BoDa6WyA3b7XYgZpZAAM%2BWm-xmhxsAO2oIOEKYWQAjMlYb94VAQMQHiAUMjNo10V5JHCZiC9mjmkislskkUYOC0PQKc0esBcJh2U1MKTipVMri-oSmPEqDBUsNRuN8kiFTAnGEEE5FstVuttvQ6WDDh4qZKifESQTiNLZd4pPNAiENcEKGRYoDUhcrq8CmyxTRyoxWOwOFlnlAandDq0Ol0YJ9BqqlT8-g7teFdYCWTt6Ryjg9TtkvZlrppfQ0C08xC9cu8E98af9AUjDQyMTDG39EcCUbnDlicZ3ZoTiaSs9ymFTG6280zs2yOVzjQD%2BcUOcKBf7iMr-gBiKAAHmsx5P1kb%2B9I4BAqWEEPiEGIKHwmNgYCgACINe%2BoAALMhMYUFGwMhVEyWA%2BA-YsziySsixMS4Sx9W4uVgsNqygV53jqCZ3wAbigc98SI4iSPxYgAA8HBUYgfEg%2BCrhgx5sjaToVBrN5sPyb8oCyJpengBxcUbKUZTlIYRjGCYVQk9Vwi1JYMxovVWTJHMjUpU1h3NS0RNlIA)

**Expected Behavior**

No type errors.  Pyright passes it cleanly: [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCSMApiAIYBGANsQDSEkQDOAaksQO70AqixLZEPQCCKOAFgAUKEhQAxmCo05MJGBRMAdJTmZsufAGEySyjSlTuUALxReCfoIAUAcm4uAlBckATYsCgIOAB9CDIEYLQwMB8nYABXFDkALihjU2piAG1NXJ4AXXokEnJM1KJSM2zROEKoACpBNCZUsAoAK2IVDygAWgA%2BKCokJhgs7nzkqSgZqD8A4I5yBAcfYISkuMS5YLIUNaaW2HiEGiz0qiqcvLs6%2BpgTs5r6XM18-J6BuynJWb%2BoDa6WyA3b7XYgZpZAAM%2BWm-xmhxsAO2oIOEKYWQAjMlYb94VAQMQHiAUMjNo10V5JHCZiC9mjmkislskkUYOC0PQKc0esBcJh2U1MKTipVMri-oSmPEqDBUsNRuN8kiFTAnGEEE5FstVuttvQ6WDDh4qZKifESQTiNLZd4pPNAiENcEKGRYoDUhcrq8CmyxTRyoxWOwOFlnlAandDq0Ol0YJ9BqqlT8-g7teFdYCWTt6Ryjg9TtkvZlrppfQ0C08xC9cu8E98af9AUjDQyMTDG39EcCUbnDlicZ3ZoTiaSs9ymFTG6280zs2yOVzjQD%2BcUOcKBf7iMr-gBiKAAHmsx5P1kb%2B9I4BAqWEEPiEGIKHwmNgYCgACINe%2BoAALMhMYUFGwMhVEyWA%2BA-YsziySsixMS4Sx9W4uVgsNqygV53jqCZ3wAbigc98SI4iSPxYgAA8HBUYgfEg%2BCrhgx5sjaToVBrN5sPyb8oCyJpengBxcUbKUZTlIYRjGCYVQk9Vwi1JYMxovVWTJHMjUpU1h3NS0RNlIA)

**Actual Behavior**

`error: Argument 1 to "map" has incompatible type "Callable[[tuple[Callable[..., T], *tuple[Any, ...]]], T]"; expected "Callable[[tuple[object, ...]], T]"  [arg-type]`

**Your Environment**

- Mypy version used: 1.17.0
- Mypy command-line flags: -
- Mypy configuration options from `mypy.ini` (and other config files): -
- Python version used: 3.11 / 3.12 / 3.13 (checked only these in the playground)

**Comment by sterliakov** [comment]
Mypy does not typically (ever?) infer variadic tuple types AFAIC. Your first example passes because the inferred type of `func_and_args` is `list[tuple[Any, ...]]`, so you can do anything with that.

The MRE is

```python
from typing import Any

def ok(first: int, second: Any, *args: object) -> None:
    func_and_args: tuple[int, *tuple[Any, ...]] = (first, second, *args)
    
def fail(first: int, second: int, *args: object) -> None:
    func_and_args: tuple[int, *tuple[Any, ...]] = (first, second, *args) # E: \
    # Incompatible types in assignment (expression has type "tuple[object, ...]", variable has type "tuple[int, *tuple[Any, ...]]")  [assignment]
```

[playground](https://mypy-play.net/?mypy=master&python=3.12&gist=fa2254ccf12cfe989299558035726b4d)

Looks like there's some omission in join where we fail to match `(it_arg, *args)` against `*tuple[Any, ...]` part of the type, give up and use a variadic tuple of join of all elements (which is obviously `object`).

`MappingView` is not special, nested tuples aren't either - the only important thing here is that `Any` is contagious, so we end up with a tuple of `Any`s.

**Comment by sterliakov** [comment]
Sorry, my bad, this happens much earlier than any joins, right in checker. Not a bug; the snippet passes cleanly with `--enable-incomplete-feature=PreciseTupleTypes`:

https://mypy-play.net/?mypy=master&python=3.12&flags=strict&enable-incomplete-feature=PreciseTupleTypes&gist=a83b792c24c3a282e4617accb5a38035
